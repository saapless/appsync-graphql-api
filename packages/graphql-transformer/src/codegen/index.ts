export * from "./ast";
export * from "./printer";
export * from "./document";
