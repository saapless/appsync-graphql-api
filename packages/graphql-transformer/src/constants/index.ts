export const ConnectionDirective = Object.freeze({
  HAS_ONE: "hasOne",
  HAS_MANY: "hasMany",
});

export const ModelDirective = Object.freeze({
  MODEL: "model",
});

export const UtilityDirective = Object.freeze({
  SERVER_ONLY: "serverOnly",
  CLIENT_ONLY: "clientOnly",
  READ_ONLY: "readOnly",
  WRITE_ONLY: "writeOnly",
  FILTER_ONLY: "filterOnly",
  INTERNAL: "internal",
});

export const ScalarType = Object.freeze({
  STRING: "String",
  INT: "Int",
  FLOAT: "Float",
  BOOLEAN: "Boolean",
  ID: "ID",
  LONG: "Long",
});

export const RelationType = Object.freeze({
  ONE_TO_ONE: "oneToOne",
  ONE_TO_MANY: "oneToMany",
  MANY_TO_MANY: "manyToMany",
});
