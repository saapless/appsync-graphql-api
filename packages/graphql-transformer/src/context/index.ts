export * from "./TransformerContext";
export type {
  DescriptorOperation,
  Key,
  KeyOperator,
  KeyValue,
  OperationActionType,
  ResolverDescriptor,
  ReturnType,
} from "./ResolverLoader";
