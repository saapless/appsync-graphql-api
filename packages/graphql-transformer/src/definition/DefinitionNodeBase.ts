import { ConstDirectiveNode } from "graphql";
import { DirectiveNode } from "./DirectiveNode";

export class WithDirectivesNode {
  name: string;
  directives?: DirectiveNode[] | undefined;

  constructor(name: string, directives?: DirectiveNode[]) {
    this.name = name;
    this.directives = directives;
  }

  public hasDirective(name: string) {
    return this.directives?.some((directive) => directive.name === name) ?? false;
  }

  public getDirective(name: string) {
    return this.directives?.find((directive) => directive.name === name);
  }

  public addDirective(directive: string | DirectiveNode | ConstDirectiveNode) {
    const node =
      directive instanceof DirectiveNode
        ? directive
        : typeof directive === "string"
          ? DirectiveNode.create(directive)
          : DirectiveNode.fromDefinition(directive);

    if (this.hasDirective(node.name)) {
      throw new Error(`Directive ${node.name} already exists on node ${this.name}`);
    }

    this.directives = this.directives ?? [];
    this.directives.push(node);
    return this;
  }

  public removeDirective(name: string) {
    this.directives = this.directives?.filter((directive) => directive.name !== name);
    return this;
  }
}

// export class WithArgumentsNode {
//   name: string;
//   arguments?: InputValueNode[] | undefined;

//   constructor(name: string, args?: InputValueNode[]) {
//     this.name = name;
//     this.arguments = args;
//   }

//   public hasArgument(name: string) {
//     return this.arguments?.some((arg) => arg.name === name) ?? false;
//   }

//   public getArgument(arg: string) {
//     return this.arguments?.find((argument) => argument.name === arg);
//   }

//   public addArgument(argument: InputValueNode | InputValueDefinitionNode) {
//     const node =
//       argument instanceof InputValueNode ? argument : InputValueNode.fromDefinition(argument);

//     if (this.hasArgument(node.name)) {
//       throw new Error(`Argument ${node.name} already exists on node ${this.name}`);
//     }

//     this.arguments = this.arguments ?? [];
//     this.arguments.push(node);
//     return this;
//   }

//   public removeArgument(name: string) {
//     this.arguments = this.arguments?.filter((arg) => arg.name !== name);
//     return this;
//   }
// }
