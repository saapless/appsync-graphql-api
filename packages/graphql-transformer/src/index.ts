export * from "./transformer";
export { definitionFromSource } from "./utils";
