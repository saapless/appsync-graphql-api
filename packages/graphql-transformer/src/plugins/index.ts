export { TransformerPluginBase, type IPluginFactory } from "./TransformerPluginBase";
export { AWSTypesPlugin } from "./AWSTypesPlugin";
export { AuthPlugin } from "./AuthPlugin";
export { ModelPlugin } from "./ModelPlugin";
export { NodeInterfacePlugin } from "./NodeInterfacePlugin";
export { ConnectionPlugin } from "./ConnectionPlugin";
