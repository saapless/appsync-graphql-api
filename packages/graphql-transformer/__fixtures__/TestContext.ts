import { TransformerContext, TransformerContextConfig } from "../src/context";

interface TestingTransformerConfig extends TransformerContextConfig {
  allowPrint?: boolean;
}

export class TestContext extends TransformerContext {
  public allowPrint: boolean;
  public files: Map<string, string> = new Map();

  constructor(config: TestingTransformerConfig) {
    super(config);
    this.allowPrint = config.allowPrint ?? false;
    this.files = new Map();
  }

  public override createOutputDirectory(path: string) {
    return path;
  }

  public printFile(filePath: string, content: string) {
    if (this.allowPrint) {
      return super.printFile(filePath, content);
    }

    this.files.set(filePath, content);
  }

  public printScript(filePath: string, content: string) {
    if (this.allowPrint) {
      return super.printScript(filePath, content);
    }

    this.files.set(filePath, content);
  }
}
