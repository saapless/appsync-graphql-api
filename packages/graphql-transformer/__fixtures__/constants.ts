import { DataSourceManagerConfig } from "../src/context/DataSourceManager";

export const TEST_DS_CONFIG = {
  primaryDataSourceName: "TestDataSource",
  dataSources: {
    TestDataSource: {
      type: "DYNAMO_DB",
      config: {
        tableName: "TestTable",
      },
    },
    NoneDataSource: {
      type: "NONE",
    },
  },
} satisfies DataSourceManagerConfig;
