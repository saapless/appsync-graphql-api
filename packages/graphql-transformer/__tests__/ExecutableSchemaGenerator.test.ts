import { beforeAll } from "@jest/globals";
import { TestContext } from "../__fixtures__/TestContext";
import { DocumentNode } from "../src/definition";
import { ExecutableSchemaGenerator } from "../src/plugins";

const schema = /* GraphQL */ `
  type User {
    id: ID!
    name: String!
    email: String!
    role: UserRole!
    tasks: Task
  }

  enum UserRole {
    ADMIN
    USER
  }

  type Task {
    id: ID!
    title: String!
    description: String!
    user: User!
  }

  type Query {
    viewer: User
  }
`;

const context = new TestContext({
  document: DocumentNode.fromSource(schema),
  outputDirectory: "__test__",
});

const generator = new ExecutableSchemaGenerator(context, { outDir: "executable-schema" });

describe("ExecutableSchemaGenerator", () => {
  beforeAll(() => {
    for (const node of context.document.definitions.values()) {
      generator.execute(node);
    }
  });

  it("generates executable schema", () => {
    generator.generate();
    expect(context.files.get("executable-schema/schema.ts")).toMatchSnapshot();
  });
});
