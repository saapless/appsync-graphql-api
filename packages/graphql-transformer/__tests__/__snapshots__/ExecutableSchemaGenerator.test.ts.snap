// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecutableSchemaGenerator generates executable schema 1`] = `
"/* eslint-disable */
import { GraphQLObjectType, GraphQLNonNull, GraphQLID, GraphQLString, GraphQLEnumType, GraphQLSchema } from "graphql";
let User: GraphQLObjectType<Schema.User, ResolverContext>;
let Task: GraphQLObjectType<Schema.Task, ResolverContext>;
let Query: GraphQLObjectType<undefined, ResolverContext>;
const UserRole = new GraphQLEnumType({ name: "UserRole", values: { ADMIN: { value: "ADMIN" }, USER: { value: "USER" } } });
Task = new GraphQLObjectType({ name: "Task", fields: () => ({ id: { type: new GraphQLNonNull(GraphQLID) }, title: { type: new GraphQLNonNull(GraphQLString) }, description: { type: new GraphQLNonNull(GraphQLString) }, user: { type: new GraphQLNonNull(User) } }) });
User = new GraphQLObjectType({ name: "User", fields: () => ({ id: { type: new GraphQLNonNull(GraphQLID) }, name: { type: new GraphQLNonNull(GraphQLString) }, email: { type: new GraphQLNonNull(GraphQLString) }, role: { type: new GraphQLNonNull(UserRole) }, tasks: { type: Task } }) });
Query = new GraphQLObjectType({ name: "Query", fields: () => ({ viewer: { type: User } }) });
export const schema = new GraphQLSchema({
    query: Query,
    mutation: Mutation,
    types: [User, UserRole, Task]
});
"
`;
